tls/s2n_handshake_io.c

```diff
@@ -1053,7 +1053,9 @@ int s2n_conn_set_handshake_type(struct s2n_connection *conn)
             POSIX_GUARD_RESULT(s2n_validate_ems_status(conn));
 
             /* Set up the handshake to send a session ticket since a valid ticket was not provided */
-            if (s2n_config_is_encrypt_decrypt_key_available(conn->config) == 1) {
+            bool key_available = false;
+            POSIX_GUARD_RESULT(s2n_config_is_encrypt_key_available(conn->config, &key_available));
+            if (key_available) {
```

Comment:
https://github.com/aws/s2n-tls/pull/4513#discussion_r1580235890

I was thinking this was just a minor choice since we generally ignore failures from the whole s2n_decrypt_session_ticket code path anyway (or at least should), but here you're potentially introducing a new failure that won't be ignored. That's a potential behavior change, even if a VERY small one. Should failing to check for an encrypt key really end the connection? Wouldn't we WANT to ignore that error and treat it as equivalent to not finding an encrypt key?

I could understand the change if there was a compelling reason for it, but I'm really not seeing one. We would never expect a failure while checking for a key, but if one happened, it shouldn't kill the connection.

Reply:
Sure, I think that's valid considering the behavior change. But do we still want to change the return type from int to S2N_RESULT and use it like `s2n_config_is_encrypt_key_available(conn->config) == S2N_SUCCESS` without the output parameter?

Reply:
Correct, keep the S2N_RESULT, it stops people from abusing the return value. You'd use it with s2n_result_is_ok() or s2n_result_is_error()

Reply:
I realized an output parameter is required for `s2n_config_is_encrypt_key_available()` if we want to use S2N_RESULT return type, because we don't want to return S2N_RESULT_ERROR to indicate key was not found. Without being able to indicate that key was not found, test cases that uses no keys or expired key will fail because they can't catch those cases. So the option we have is:

1. change return type back to int (people can abuse return type)
2. use output param and error if failed to search for a key (unwanted behavior change)
3. use output param, but use it inside `s2n_result_is_ok()`. (This will be an awkward option because the output param end up not being utilized)

I'm leaning toward either option 1 or 3 as they maintain the behavior. What's your opinion on this?

Reply:
maddeleine walked me through this concern and was able to find a solution to it